File: Data-Structure-I/5-Tree/P94.cpp
// Date: Wed, Oct 6, 2021
// 94. Binary Tree Inorder Traversal

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

// recursive approach
// time complexity O(n)
// space complexity: if count call stacks then in the worst case O(n); average case is O(logn)
void helper(vector<int>& ans, TreeNode* current) {
	if (current) {
		helper(ans, current->left);
		ans.push_back(current->val);
		helper(ans, current->right);
	}
}

vector<int> inorderTraversal(TreeNode* root) {
	vector<int> ans;
	helper(ans, root);
	return ans;
}

// iterative approach using stack
// time complexity O(n)
// space complexity O(n)
vector<int> inorderTraversal(TreeNode* root) {
	vector<int> ans;
	stack<TreeNode*> nodes;
	TreeNode* cur = root;
	while(cur || !nodes.empty()) {
		while (cur) {
			nodes.push(cur);
			cur = cur->left;
		}
		cur = nodes.top();
		nodes.pop();
		ans.push_back(cur->val);
		cur = cur->right;
	}
	return ans;
}

File: Others/P94.java
package others;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;

/**
 * Date: April 18, 2022 - Redo: Sept 07, 2022
 * 94. Binary Tree Inorder Traversal
 * https://leetcode.com/problems/binary-tree-inorder-traversal/solution/
 */

public class P94 {
    // 1 - iterative inorder traversal
    // time & space: O(n)
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> ans = new ArrayList<>();
        Deque<TreeNode> stack = new ArrayDeque<>();
        TreeNode current = root;
        while (current != null || !stack.isEmpty()) {
            while (current != null) {
                stack.push(current);
                current = current.left;
            }
            current = stack.pop();
            ans.add(current.val);
            current = current.right;
        }
        return ans;
    }

    // 2 - Morris Traversal
    // time: O(n)
    // space: O(1)
    public List<Integer> _inorderTraversal(TreeNode root) {
        List<Integer> ans = new ArrayList<>();
        TreeNode current = root;
        TreeNode pred;
        while (current != null) {
            if (current.left != null) {
                // find the most-right node in the left subtree of the current node
                pred = current.left;
                while (pred.right != null) pred = pred.right;
                // make the current node the right child of the right-most node
                pred.right = current;
                // erase the left pointer of the current node
                TreeNode tmp = current.left;
                current.left = null;
                // iterate to the tmp node
                current = tmp;
            } else {
                ans.add(current.val);
                current = current.right;
            }
        }
        return ans;
    }

    // 3 - Recursive approach (Redo on Sept 07, 2022)
    private List<Integer> ans;

    public List<Integer> __inorderTraversal(TreeNode root) {
        ans = new ArrayList();
        inorder(root);
        return ans;
    }

    private void inorder(TreeNode node) {
        if (node != null) {
            inorder(node.left);
            ans.add(node.val);
            inorder(node.right);
        }
    }

    // Definition for a binary tree node.
    private class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
}

/**
 * reference
 * https://leetcode.com/problems/binary-tree-inorder-traversal/solution/
 */